### ISSUE TITLE
Cannot add AdditionalFiles in custom targets file

### ISSUE THREAD

**Issue by sharwell** [dotnet/roslyn#5999]
I'm working to resolve issues surrounding the use of **stylecop.json** as an `AdditionalFiles` item used by our analyzers. For additional background on this work, see DotNetAnalyzers/StyleCopAnalyzers#1581.

I created the following **StyleCop.Analyzers.targets** file and imported in as part of our project:

``` xml
<?xml version="1.0" encoding="utf-8"?>
<Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
  <Target Name="InjectStyleCopSettings"
          BeforeTargets="CoreCompile">
    <ItemGroup>
      <StyleCopSettings Include="@(None)"
                        Condition="'%(Filename)%(Extension)' == 'stylecop.json'"/>
      <AdditionalFiles Include="%(StyleCopSettings.Identity)" />
    </ItemGroup>
  </Target>
</Project>
```

In the main C# project file, I have the following item:

``` xml
<None Include="..\stylecop.json">
  <Link>stylecop.json</Link>
</None>
```
### Description of problem

When the project is compiled in the IDE or on the command line, no errors/warnings are reported from the StyleCop.Analyzers analyzers. However, if you open the project in Visual Studio and wait for a bit, you'll see over 100 errors appear because the analyzers are using default settings instead of our customized **stylecop.json** settings file. This suggests that the target `InjectStyleCopSettings` which I defined above is executing during regular builds, but is not executing as part of the IntelliSense process.

I noticed that if I place the `InjectStyleCopSettings` target directly inside the C# project file instead of including it as a separate **.targets** file, both the build and the IntelliSense features work properly.

:question: Is there a way to use a separate targets file and still have both the IntelliSense and build work properly?

**Comment by sharwell** [comment]
This appears to be caused by caching of some form. When I first created the **.targets** file, it contained an error which caused it to not add the correct files to `AdditionalFiles`. I then modified the file to the form above (which works properly for a build).

:exclamation: Note that restarting Visual Studio **does not** force the IntelliSense engine to reload the updated **.targets** files. If I modify the file to break it again, the IntelliSense engine continues to work properly for both diagnostics and code fixes for an unknown period of time, even though the build itself will fail with warnings or errors.

The specific way I changed the targets file to exhibit this behavior is by using the following line:

```xml
<AdditionalFiles Include="@(%(StyleCopSettings.Identity))" />
```

**Comment by Pilchie** [comment]
This is a known limitation of either the project system or msbuild itself.  We're not sure which, but @davkean and I were just talking about it yesterday.  This is something we hope to address in the next major VS release as we move to a new project system based on CPS.

For the closing and re-opening, we cache the results of the IntelliSense build in the .suo, which is likely the second issue you were seeing.

**Comment by Pilchie** [comment]
@davkean and/or @srivatsn Can you move this to the project system repo?

**Comment by srivatsn** [comment]
The core issue is tracked by https://github.com/dotnet/roslyn-project-system/issues/268
